workflow-api/src/lib.rs

```diff
@@ -0,0 +1,205 @@
+    }
+}
+// ... also convenience functions for constructing C-A-N, etc.
+
+/// A workflow context which contains only information, but does not allow any commands to
+/// be created.
+pub struct SafeWfContext {
+    // TODO
+}
+
+/// TODO: Placeholder, exists in SDK and would move into this crate & (likely) become a trait
+pub struct WfContext {}
+impl WfContext {
+    pub async fn timer(&self, _: Duration) {
+        todo!()
+    }
+}
+
+pub struct SignalDefinition<WF: Workflow> {
+    // TODO: Could be a matching predicate, to allow for dynamic registration
+    name: String,
+    // The handler input type must be erased here, since otherwise we couldn't store/return the
+    // heterogeneous collection of definition types in the workflow itself. The signal macro
+    // will wrap the user's function with code that performs deserialization.
+    handler: Box<dyn FnMut(&mut WF, Payload) -> Result<(), anyhow::Error>>,
+}
+pub struct QueryDefinition<WF: Workflow> {
+    // TODO: Could be a matching predicate, to allow for dynamic registration
+    name: String,
+    // The query macro will wrap the user's function with code that performs deserialization of
+    // input and serialization of output, as well as error boxing.
+    handler: Box<dyn FnMut(&WF, Payload) -> Result<Payload, anyhow::Error>>,
+}
+
+/// TODO: Placeholders, likely belong inside protos crate. These will be auto-implemented for
+///   anything using serde already (which I expect is how virtually everyone will do this).
+pub trait TemporalSerializable {}
+impl<T> TemporalSerializable for T {}
+pub trait TemporalDeserializable {}
+impl<T> TemporalDeserializable for T {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::FutureExt;
+    use std::{collections::HashMap, marker::PhantomData};
+
+    // Workflow implementation example
+    struct MyWorkflow {
+        foo: u64,
+        bar: HashMap<String, u64>,
+    }
+
+    impl Workflow for MyWorkflow {
+        type Input = String;
+        type Output = u64;
+        const NAME: &'static str = "MyWorkflowType";
+
+        fn new(input: Self::Input, _ctx: SafeWfContext) -> Self {
+            let mut bar = HashMap::new();
+            bar.insert(input, 10);
+            Self { foo: 0, bar }
+        }
+
+        fn run(
+            &mut self,
+            ctx: WfContext,
+        ) -> BoxFuture<Result<WfExitValue<Self::Output>, anyhow::Error>> {
+            async move {
+                ctx.timer(Duration::from_secs(1)).await;
+                self.foo = 1;
+                // See activity definitions file
+                ctx.my_act_fn("Hi!").await.unwrap();
+                // The into() is unfortunately unavoidable without making C-A-N and confirm cancel
+                // be errors instead. Personally, I don't love that and I think it's not idiomatic
+                // Rust, whereas needing to `into()` something is. Other way would be macros, but
+                // it's slightly too much magic I think.
```

Comment:
Concerning the `into()`, I wonder if maybe instead of `Result<WfExitValue<Self::Output>, anyhow::Error>` you have `Result<Self::Output, TemporalNonSuccess>` of some form. That way `Ok` _only_ represents literal success, not cancellation or continue as new.

I think it's worth optimizing for the normal/happy path of `Ok` being the literal result and all else being the `Err`.

Reply:
The problem is CAN _is_ a form of success. The semantics of it are my real issue. `TemporalNonSuccess` as a name is a perfect example. CAN isn't "NonSuccess".

Now, to be fair, currently I have WillRespondAsync as an error in activities. My rationale is that's less common than CAN. I get optimizing for the happy path, but `into` is easy and the compiler error will make it super obvious if the user hasn't done it.

I could live with it, for sure, but it just doesn't feel correct to me or particularly idiomatic for Rust

Reply:
Since continue as new has to be invoked top level to use this type anyways, I wonder if you can make it a macro (so they don't have to `?` or `Err` it), e.g. `continue_as_new!()`. That can kinda hide it. Otherwise, I agree that continue as new and cancel aren't technically errors, it's just a bit rough on the simple path. I can see both arguments here.